src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +52,356 @@ GlobalShortcutWin::~GlobalShortcutWin() {
+			return;
+		}
+	}
 
-	// Convert the low-level key event to
-	// a DirectInput key ID.
-	unsigned int keyid = static_cast< unsigned int >((scancode << 8) | 0x4);
-	if (extended) {
-		keyid |= 0x8000U;
+	auto &device = (*iter).second;
+#ifdef USE_XBOXINPUT
+	if (device.xinput && m_xinput) {
+		return;
 	}
+#endif
+	auto data = reinterpret_cast< PHIDP_PREPARSED_DATA >(&device.data[0]);
 
-	// NumLock and Pause need special handlinGlobal::get().
-	// For those keys, the method above of setting
-	// bit 15 high when the LLKHF_EXTENDED flag is
-	// set on the low-level key event does not work.
-	//
-	// When we receive a low-level Windows
-	// Pause key event, the extended flag isn't
-	// set, but DirectInput expects it to be.
-	//
-	// The opposite is true for NumLock key,
-	// where the extended flag for the low-level
-	// Windows event is set, but DirectInput expects
-	// it not to be.
-	//
-	// Without this fix-up, we would emit Pause as
-	// NumLock, and NumLock as pause. That was
-	// problematic, because at the same time,
-	// DirectInput would emit the correct key.
-	// This meant that when pressing one of Pause
-	// and NumLock, shortcut actions for both keys
-	// would be triggered.
-	//
-	// Originally reported in mumble-voip/mumble#1353
-	if (vkcode == VK_PAUSE) {
-		// Always set the extended bit for Pause.
-		keyid |= 0x8000U;
-	} else if (vkcode == VK_NUMLOCK) {
-		// Never set the extended bit for NumLock.
-		keyid &= ~0x8000U;
+	ULONG nUsages = device.buttons.size();
+	std::vector< USAGE > usages(nUsages);
+	if (HidP_GetUsages(HidP_Input, HID_USAGE_PAGE_BUTTON, 0, &usages[0], &nUsages, data, &reports[0], reportSize)
+		!= HIDP_STATUS_SUCCESS) {
+		return;
 	}
 
-	ql << keyid;
-	ql << QVariant(QUuid(GUID_SysKeyboard));
-	bool suppress = gsw->handleButton(ql, down);
+	device.buttons.assign(device.buttons.size(), false);
+	for (ULONG i = 0; i < nUsages; ++i) {
+		device.buttons[usages[i] - device.usageRange.first] = true;
+	}
 
-	return suppress;
+	for (size_t i = 0; i < device.buttons.size(); ++i) {
+		const QList< QVariant > ql{ QVariant::fromValue(DeviceTypes::HID), QString::fromStdString(device.name),
+									QString::fromStdString(device.prefix), i };
+		handleButton(ql, device.buttons[i]);
+	}
 }
 
-bool GlobalShortcutWin::handleMouseMessage(unsigned int btn, bool down) {
-	GlobalShortcutWin *gsw = static_cast< GlobalShortcutWin * >(engine);
-
-	bool suppress = false;
+void GlobalShortcutWin::on_keyboardMessage(const uint16_t flags, uint16_t scanCode, const uint16_t virtualKey) {
+	const auto up = flags & RI_KEY_BREAK;
+	const auto e0 = flags & RI_KEY_E0;
+	const auto e1 = flags & RI_KEY_E1;
 
-	if (btn > 0) {
-		QList< QVariant > ql;
-		ql << static_cast< unsigned int >((btn << 8) | 0x4);
-		ql << QVariant(QUuid(GUID_SysMouse));
+	if (virtualKey == VK_NUMLOCK) {
+		// Correct PAUSE/BREAK and NUM LOCK silliness, and set the extended bit.
+		scanCode = (MapVirtualKey(virtualKey, MAPVK_VK_TO_VSC) | 0x100);
```

Comment:
What kind of silliness is that and how does the fix work? (-> should be part of this comment)

Reply:
From https://blog.molecular-matters.com/2011/09/05/properly-handling-keyboard-input:

> Keys like “Pause/Break” and “Numlock” act strangely, sometimes like they are not even the same physical key – they use so-called [escaped sequences](http://www.win.tue.nl/~aeb/linux/kbd/scancodes-1.html) which we have to decipher.